### DISCUSSION TITLE
Validation questions

### DISCUSSION THREAD

**Discussion by deaddr0p** [FamilySearch/GEDCOM#260] [General]
I'm working on a golang validator for gedcom 7 files. Mostly, I'm doing this as an exercise for my own benefit but I would like to add some cli features down the road to manipulate the various gedcom files I have laying around.

I plan to ingest the `extracted-files` folder to implement the validation rules. This will let me update the validation rules any time a new gedcom version is released. I currently parse the yaml files in the `tags` directory. I'm wondering is there is any additional information from the TSV files that I would also require to properly validate a file.

I'm also curious why HEAD are TRLR are defined as `pseudostructure` and don't have yaml files. I get that they are something of a special case because they can only occur once and must be the first/last records in a gedcom file. I am handling them with small hacks to load the abnf rules and allowed substructures for HEAD but it would be nice if this could also be automatic when I parse the contents of `extracted-files`.

**Comment by deaddr0p** [comment]
Dug in a bit and decided I didn't really need anything from the `tsv` files.

I published a new repository with an early version of what I'm working on. Still a lot of work to do but it currently does a few things.

- Reads a v7 GEDCOM file and loads it into a structure that can be used for further work.
- Writes a v7 GECOM file that process the loaded document. Reading and writing creates an exact copy of the original file.
- Reads the yaml files, imports the ABNF grammars for the current specification and dynamically creates validation functions. I still need to write code to walk the tree for each record so I can properly test the validators.

If anyone is interested in taking a peek, it's at https://github.com/funwithbots/go-gedcom

**Reply by clarkegj** [reply]
We need another parser.  Is go-gedcom ready to be listed at:
https://gedcom.io/tools/

You or I can add it if you like.

Gordon
GEDCOM.io manager

**Comment by clarkegj** [comment]
Is this ready to add to Tools for FamilySearch GEDCOM - FamilySearch GEDCOM<https://gedcom.io/tools/> ?

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

From: Bill ***@***.***>
Sent: Saturday, January 7, 2023 4:42 PM
To: ***@***.***>
Cc: ***@***.***>
Subject: [Ext:] Re: [FamilySearch/GEDCOM] Validation questions (Discussion #260)

[External Email]

Dug in a bit and decided I didn't really need anything from the tsv files.

I published a new repository with an early version of what I'm working on. Still a lot of work to do but it currently

  *   Reads a v7 GEDCOM file and loads it into a structure that can be used for further work.
  *   Writes a v7 GECOM file that process the loaded document. Reading and writing creates an exact copy of the original file.
  *   Reads the yaml files, imports the ABNF grammars for the current specification and dynamically creates validation functions. I still need to write code to walk the tree for each record so I can properly test the validators.

If anyone is interested in taking a peek, it's at https://github.com/funwithbots/go-gedcom<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ffunwithbots%2Fgo-gedcom&data=05%7C01%7Cgordon%40familysearch.org%7C9776a73f16b049fdee4908daf108dd96%7C61e6eeb35fd74aaaae3c61e8deb09b79%7C0%7C0%7C638087317636190557%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=PblNYAl%2B8G6Ci90ROcUadpx5ChcpMiIeGwJJM5fFnCI%3D&reserved=0>

**Reply by deaddr0p** [reply]
Hi Gordon,

The software is a bit fresh but based on the testing I've done, I feel like it could be listed. More functionality is still being added but it does indeed parsed GEDCOM 7 files and makes them usable in golang programs.

Bill

**Comment by clarkegj** [comment]
OK.  Please join https:/github.com/familysearch/gedcom.io so that you can see and comment on the change I make to the website.

Gordon
[https://opengraph.githubassets.com/4f36924539c6beb92bef7066a4c6aa568987e2624c6747c354249d5e6de8f4a8/FamilySearch/GEDCOM.io]<https://github.com/familysearch/gedcom.io>
GitHub - FamilySearch/GEDCOM.io: Files for the GEDCOM.io website<https://github.com/familysearch/gedcom.io>
Files for the GEDCOM.io website. Contribute to FamilySearch/GEDCOM.io development by creating an account on GitHub.
github.com

From: Bill Shaw ***@***.***>
Sent: Monday, January 9, 2023 3:58 PM
To: FamilySearch/GEDCOM ***@***.***>
Cc: Gordon Clarke ***@***.***>; Comment ***@***.***>
Subject: [Ext:] Re: [FamilySearch/GEDCOM] Validation questions (Discussion #260)

[External Email]

Hi Gordon,

The software is a bit fresh but based on the testing I've done, I feel like it could be listed. More functionality is still being added but it does indeed parsed GEDCOM 7 files and makes them usable in golang programs.

Bill

**Comment by deaddr0p** [comment]
Gordon,

I opened a pull request and tagged you with the change. 
https://github.com/FamilySearch/GEDCOM.io/pull/64

Bill

**Comment by clarkegj** [comment]
Bill,

Thank you for your work.  I have merged you changes to the Tools page so it is live with your link to go-gedcom .

Gordon

From: Bill Shaw ***@***.***>
Sent: Tuesday, January 10, 2023 5:27 AM
To: FamilySearch/GEDCOM ***@***.***>
Cc: Gordon Clarke ***@***.***>; Comment ***@***.***>
Subject: [Ext:] Re: [FamilySearch/GEDCOM] Validation questions (Discussion #260)

[External Email]

Gordon,

I opened a pull request and tagged you with the change.
FamilySearch/GEDCOM.io#64<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FFamilySearch%2FGEDCOM.io%2Fpull%2F64&data=05%7C01%7Cgordon%40familysearch.org%7Cc0f6cd61dfbf43e7265508daf3060170%7C61e6eeb35fd74aaaae3c61e8deb09b79%7C0%7C0%7C638089504366166784%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=yDjopMjiEu3lZplf05KHl8ijG1aUE0bBxOTGpL%2FJen0%3D&reserved=0>

Bill

-

**Comment by tychonievich** [comment]
> I'm working on a golang validator for gedcom 7 files. Mostly, I'm doing this as an exercise for my own benefit but I would like to add some cli features down the road to manipulate the various gedcom files I have laying around.
> 
> I'm wondering is there is any additional information from the TSV files that I would also require to properly validate a file.

That information wasn't previously available, but I've now added a [README in the extracted-files directory](https://github.com/FamilySearch/GEDCOM/tree/main/extracted-files) that hopefully provides the information you wanted. If there is something missing there, please let me know.

> I'm also curious why HEAD are TRLR are defined as `pseudostructure`

That's my doing. When drafting chapter 1 of the 7.0.0 spec, HEAD, TRLR, and CONT kept being exceptions to rules (such as reordering rules, Xref identifier rules, and extension substructure rules) and it was easier to just say "these three are special" than worry that I'd overlooked some place where they were special. The rest of the 7.0.0 drafting team pushed back on this, as it seemed strange to have this "pseudostructure" idea, but the alternative was putting "except HEAD, TRLR, and CONT" in various places throughout that chapter and that seemed less readable, so the pseudostructure idea made it into 7.0.

>  I'm also curious why HEAD are TRLR […] don't have yaml files. […] it would be nice if this could also be automatic when I parse the contents of `extracted-files`.

A good question. I've created #263 to track this. Technically the substructures of HEAD are present in the extracted_files (e.g. <https://gedcom.io/terms/v7/GEDC> has a `superstructures` field with `"HEAD pseudostructure": "{1:1}"`) but I can see how having a YAML file for the header itself could help.
